redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+}
+
+// Implement the `ClusterInScan` trait for `InnerCore` of async cluster connection.
+#[async_trait]
+impl<C> ClusterInScan for Core<C>
+where
+    C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
+{
+    async fn get_address_by_slot(&self, slot: u16) -> RedisResult<String> {
+        let address = self
+            .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
+            .await;
+        match address {
+            Some(addr) => Ok(addr),
+            None => {
+                if self.is_all_slots_covered().await {
+                    Err(RedisError::from((
+                        ErrorKind::IoError,
+                        "Failed to get connection to the node cover the slot, please check the cluster configuration ",
+                    )))
+                } else {
+                    Err(RedisError::from((
+                        ErrorKind::NotAllSlotsCovered,
+                        "All slots are not covered by the cluster, please check the cluster configuration ",
+                    )))
+                }
+            }
+        }
+    }
+
+    async fn get_address_epoch(&self, address: &str) -> Result<u64, RedisError> {
+        self.as_ref().get_address_epoch(address).await
+    }
+    async fn get_slots_of_address(&self, address: &str) -> Vec<u16> {
+        self.as_ref().get_slots_of_address(address).await
+    }
+    // Refresh the topology of the cluster
+    async fn refresh_slots(&self) -> RedisResult<()> {
+        ClusterConnInner::refresh_slots_with_retries(self.to_owned()).await
+    }
+    async fn route_command(&self, cmd: &Cmd, address: &str) -> RedisResult<Value> {
+        let core = self.to_owned();
+        InnerCore::route_command_inner(core, cmd.clone(), address).await
+    }
+    async fn is_all_slots_covered(&self) -> bool {
+        self.all_slots_covered().await
+    }
+}
+
+/// Perform a cluster scan operation.
+/// This function performs a scan operation in a Redis cluster using the given `ClusterInScan` connection.
+/// It scans the cluster for keys based on the given `ClusterScanArgs` arguments.
+/// The function returns a tuple containing the new scan state cursor and the keys found in the scan operation.
+/// If the scan operation fails, an error is returned.
+///
+/// # Arguments
+/// * `core` - The connection to the Redis cluster.
+/// * `cluster_scan_args` - The arguments for the cluster scan operation.
+///
+/// # Returns
+/// A tuple containing the new scan state cursor and the keys found in the scan operation.
+/// If the scan operation fails, an error is returned.
+pub(crate) async fn cluster_scan<C>(
+    core: C,
+    cluster_scan_args: ClusterScanArgs,
+) -> RedisResult<(ScanStateCursor, Vec<Value>)>
+where
+    C: ClusterInScan,
+{
+    let ClusterScanArgs {
+        scan_state_cursor,
+        match_pattern,
+        count,
+        object_type,
+    } = cluster_scan_args;
+    // If scan_state is None, meaning we start a new scan
+    let mut scan_state;
```

Comment:
No need to declare it as `mut`.

Instead use this:

```rust
let scan_state =  match scan_state_cursor.get_state_from_wrraper() {
        Some(state) => {
            state,
        }
        None => match ScanState::initiate_scan(&core).await {
            Ok(state) => {
                state,
            }
            Err(err) => {
                return Err(err);
            }
        },
    };
```

Reply:
later in the function theres another mutation